Penumbra/UI/AdvancedWindow/ModEditWindow.Models.cs

```diff
@@ -10,26 +11,130 @@ public partial class ModEditWindow
+        if (_submeshAttributeTagWidgets.Count != submeshTotal)
+        {
+            _submeshAttributeTagWidgets.Clear();
+            _submeshAttributeTagWidgets.AddRange(
+                Enumerable.Range(0, submeshTotal).Select(_ => new TagButtons())
+            );
+        }
+
         var ret = false;
-        for (var i = 0; i < file.Materials.Length; ++i)
+
+        for (var i = 0; i < file.Meshes.Length; ++i)
+            ret |= DrawMeshDetails(file, i, disabled);
+
+        ret |= DrawOtherModelDetails(file, disabled);
+
+        return !disabled && ret;
+    }
+
+    private static bool DrawMeshDetails(MdlFile file, int meshIndex, bool disabled)
+    {
+        if (!ImGui.CollapsingHeader($"Mesh {meshIndex}"))
+            return false;
+
+        using var id = ImRaii.PushId(meshIndex); 
+
+        var mesh = file.Meshes[meshIndex];
+
+        var ret = false;
+
+        // Mesh material.
+        var temp = file.Materials[mesh.MaterialIndex];
+        if (
+            ImGui.InputText("Material", ref temp, Utf8GamePath.MaxGamePathLength, disabled ? ImGuiInputTextFlags.ReadOnly : ImGuiInputTextFlags.None)
+            && temp.Length > 0
+            && temp != file.Materials[mesh.MaterialIndex]
+        ) {
+            file.Materials[mesh.MaterialIndex] = temp;
+            ret = true;
+        }
+
+        // Submeshes.
+        for (var submeshOffset = 0; submeshOffset < mesh.SubMeshCount; submeshOffset++)
+            ret |= DrawSubMeshDetails(file, mesh.SubMeshIndex + submeshOffset, disabled);
+
+        return ret;
+    }
+
+    private static bool DrawSubMeshDetails(MdlFile file, int submeshIndex, bool disabled)
+    {
+        using var id = ImRaii.PushId(submeshIndex);
+
+        var submesh = file.SubMeshes[submeshIndex];
+        var widget = _submeshAttributeTagWidgets[submeshIndex];
+
+        var attributes = HydrateAttributes(file, submesh.AttributeIndexMask).ToArray();
+
+        UiHelpers.DefaultLineSpace();
+        var tagIndex = widget.Draw($"Submesh {submeshIndex} Attributes", "", attributes, out var editedAttribute, !disabled);
+        if (tagIndex >= 0)
         {
-            using var id  = ImRaii.PushId(i);
-            var       tmp = file.Materials[i];
-            if (ImGui.InputText(string.Empty, ref tmp, Utf8GamePath.MaxGamePathLength,
-                    disabled ? ImGuiInputTextFlags.ReadOnly : ImGuiInputTextFlags.None)
-             && tmp.Length > 0
-             && tmp != file.Materials[i])
+            EditSubmeshAttribute(
+                file,
+                submeshIndex,
+                tagIndex < attributes.Length ? attributes[tagIndex] : null,
+                editedAttribute != "" ? editedAttribute : null
+            );
+
+            return true;
+        }
+
+        return false;
+    }
+
+    private static void EditSubmeshAttribute(MdlFile file, int changedSubmeshIndex, string? old, string? new_)
```

Comment:
this hydrate -> edit -> serialize is clunky, but IMO a lot _less_ clunky than doing a heap of bitshifting across all the submeshes in the model to realign the masks.